### ISSUE TITLE
Index signature incorrectly treats object properties

### ISSUE THREAD

**Issue by hinell** [Microsoft/TypeScript#13421]
**TypeScript Version:**  2.1.4 - Windows version `tsc.cmd`
**Description:** Index signature bug. 
```ts
var foo: { bar?: 0; [key: string]: any } = {};
foo.newProp   = '';  // Error: Property 'newProp' does not exist on type of foo
foo['newProp']= ''   // NO error!!! Same code!
```
How is it possible? :confused:  

**Expected behavior:** NO error. 
**Actual behavior:**     Error exists.

**Comment by wallverb** [comment]
FYI https://github.com/Microsoft/TypeScript/issues/12596

**Comment by hinell** [comment]
@wallverb Ah, thanks.
